### PR TITLE
wasm-encoder: remove iterator-based encoders.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR removes the iterator-based encoding functions from `wasm-encoder`
in favor of implementing `Encode` on primitive types.

### CODE REVIEW

**Review by alexcrichton — APPROVED** [review]

**Comment by peterhuene** [comment]
I'll rebase and merge this after we merge the other approved PRs just to make sure the added test coverage from those changes does not break.

**Pull request merged**
